spec/compiler/semantic/exception_spec.cr

```diff
@@ -686,4 +686,84 @@ describe "Semantic: exception" do
       end
       CR
   end
+
+  it "correctly types variable assigned inside nested exception handler (#9769)" do
+    assert_type(%(
+      int = 1
+      begin
+        begin
+          int = "a"
+        rescue
+        end
+      rescue
+      end
+      int
+    )) { union_of(int32, string) }
+  end
+
+  it "types a var after begin rescue as having all possible types and nil in begin if read (2)" do
+    assert_type(%(
+      begin
+        a = 2
+        a = 'a'
+      rescue
+      end
+      a
+      )) { union_of [int32, char, nil_type] of Type }
+  end
+
+  it "types a var after begin rescue as having all possible types in begin and rescue" do
+    assert_type(%(
+      a = 1.5
+      begin
+        a = 2
+        a = 'a'
+        a = "hello"
+      rescue ex
+        a = false
+      end
+      a
+      )) { union_of [float64, int32, char, string, bool] of Type }
```

Comment:
This looks a bit odd. I realize this spec is not new. But could the type be more restricted?
After the `begin/end` it can really only be `String` (last type from `begin` branch) or `Bool` (last type from `rescue` branch).
Just mentioning this. It's probably a separate issue.

Reply:
Actually this is quite tricky to change. For example:

```crystal
a = 2
raise "OH NO!" if something
a = 'a'
```

So it's of course not just about them being simple assignments. It's about knowing that an exception could be thrown somewhere in the middle. It's a pretty complex change, and I'm not actually sure it's either doable or worth doing. Nobody complained about this so far, so I think we shouldn't change this.